pkg/deploy/service.go

```diff
@@ -307,3 +283,103 @@ func contains(s []string, searchterm string) bool {
 	i := sort.SearchStrings(s, searchterm)
 	return i < len(s) && s[i] == searchterm
 }
+
+func buildStrategy(configStrategies map[string]model.Strategy, strategyName string) (*de.PipelinePipelineStrategy, error) {
+	strategy := configStrategies[strategyName]
+	if strategy.Canary != nil {
+		steps, err := CreateDeploymentCanaryStep(strategy)
+		if err != nil {
+			return nil, err
+		}
+		return &de.PipelinePipelineStrategy{
+			Canary: &de.KubernetesV2CanaryStrategy{
+				Steps: steps,
+			},
+		}, nil
+	} else if strategy.BlueGreen != nil {
+		redirectTrafficAfter, err := createRedirectTrafficAfter(strategy)
+		if err != nil {
+			return nil, err
+		}
+		shutdownOldVersionAfter, err := createShutdownOldVersionAfter(strategy)
+		if err != nil {
+			return nil, err
+		}
+		return &de.PipelinePipelineStrategy{
+			BlueGreen: &de.KubernetesV2BlueGreenStrategy{
+				RedirectTrafficAfter: redirectTrafficAfter,
+				ShutdownOldVersionAfter: shutdownOldVersionAfter,
+				ActiveService: strategy.BlueGreen.ActiveService,
+				PreviewService: strategy.BlueGreen.PreviewService,
+				ActiveUrl: &strategy.BlueGreen.ActiveRootUrl,
+				PreviewUrl: &strategy.BlueGreen.PreviewRootUrl,
+			},
+		}, nil
+	}
+
+	return nil, fmt.Errorf("%s is not a valid strategy; define canary or blue-green strategy", strategyName)
+}
+
+func createRedirectTrafficAfter(strategy model.Strategy) (*de.KubernetesV2RedirectTrafficAfter, error) {
+	steps := make([]de.KubernetesV2BlueGreenStep, 0, len(*strategy.BlueGreen.RedirectTrafficAfter.Steps))
+	for _, step := range *strategy.BlueGreen.RedirectTrafficAfter.Steps {
+		unit, err := createTimeUnit(step.Pause)
+		if err != nil {
+			return nil, err
+		}
+		// TODO(cat): analysis steps
+		if step.Pause != nil {
+			steps = append(
+				steps,
+				de.KubernetesV2BlueGreenStep{
+					Pause: &de.KubernetesV2PauseStep{
```

Comment:
I'm not sure if some validation has already happened here, but I would want to be sure that the user hasn't supplied an invalid step here:

```yaml
  blue-green:
    shutdownOldVersionAfter:
      steps:
        - pause:
            untilApproved: true
            unit: minutes
            duration: 5
```

^^^ this is an invalid step.

Reply:
The CLI depends on DE for validation, but agree it would be good to add some here before it tries to reach out. Especially since we can have a more user friendly message from the CLI.